src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -6708,22 +6708,32 @@ bool canBeNullableType()
                     case SyntaxKind.OpenBracketToken:
                         // Now check for arrays.
                         {
-                            var ranks = _pool.Allocate<ArrayRankSpecifierSyntax>();
-                            try
+                            // We don't want to consume a sized rank specifier in patterns, we leave it out to parse it as a length pattern.
+                            bool inPattern = (nameOptions & NameOptions.PossiblePattern) != 0;
+                            if (!inPattern || isOmittedSize())
```

Comment:
nit: Consider extracting a `isLengthPattern` local function to factor `inPattern` and the logic from `isOmittedSize` instead.
`if (!isLengthPattern()) { ... `

Reply:
That's possible at the cost of checking nameOptions every time. I even considered duplicating the loop to not check inPattern in every iteration because I doubt the rumtime can optimize it, but I decided to just keep it simple.